[FIX] SiriusAdapter: negative mode feature_only

oliveralka opened the pull request:
Introduces further changes for [REFACTOR] #3877.

[FIX] Will assume feature charge +1 or -1 for features with one masstrace based on acquisition mode. 
Additional feature charge will be set to +1 or -1 instead of 1 depending on the mode. This is used in the .ms file and later in Sirius - to use the appropriate adducts.

[TEST] Added tests for negative mode

oliveralka commented:
rebuild jenkins

timosachsenberg commented on src/openms/source/ANALYSIS/ID/SiriusMSConverter.cpp:
```diff
@@ -188,6 +188,26 @@ namespace OpenMS
         }
 
+        // set feature_charge value for msfile
+        if (p == IonSource::Polarity::POSITIVE && feature_charge == 1)
```
this is not needed

oliveralka replied:
see below.

timosachsenberg commented on src/openms/source/ANALYSIS/ID/SiriusMSConverter.cpp:
```diff
@@ -188,6 +188,26 @@ namespace OpenMS
+        if (p == IonSource::Polarity::POSITIVE && feature_charge == 0)
+        {
+          feature_charge = +1;
+          count_to_pos = count_to_pos + 1;
```
++count_to_pos ?

timosachsenberg commented on src/openms/source/ANALYSIS/ID/SiriusMSConverter.cpp:
```diff
@@ -188,6 +188,26 @@ namespace OpenMS
+        {
+          feature_charge = +1;
+        }
+        if (p == IonSource::Polarity::NEGATIVE && feature_charge == 1)
```
the whole part could be written much shorter e.g.:
if (feature_charge == 0) { feature_charge = 1; }
if (p == IonSource::Polarity::NEGATIVE) { -abs(feature_charge) }

timosachsenberg replied:
and would already handle higher charge states

oliveralka replied:
alright will change that - thank you!

hroest replied:
> and would already handle higher charge states

yes, that would be good to have -- do not assume all charges are +1 or -1 ...

oliveralka replied:
That is correct. 
In general: SIRIUS is only able to use charge +1/-1  data

hroest commented on src/openms/source/ANALYSIS/ID/SiriusMSConverter.cpp:
```diff
@@ -165,48 +165,20 @@ namespace OpenMS
+        if (p == IonSource::Polarity::POSITIVE && feature_charge == 1) { feature_charge = +1; }
+        if (p == IonSource::Polarity::NEGATIVE && feature_charge == 1) { feature_charge = -1; }
+        if (feature_charge == 0) { feature_charge = +1; ++count_to_pos; }
+        if (p == IonSource::Polarity::NEGATIVE){ -abs(feature_charge); ++count_to_neg; }
```
why not use `else if` ?

oliveralka replied:
thanks!

hroest commented on src/openms/source/ANALYSIS/ID/SiriusMSConverter.cpp:
```diff
@@ -165,48 +165,20 @@ namespace OpenMS
-          feature_charge = -1;
-          count_to_neg = count_to_neg + 1;
-        }
+        if (p == IonSource::Polarity::POSITIVE && feature_charge == 1) { feature_charge = +1; }
```
this looks like a NOP, when its 1, then set it to 1 ?

oliveralka commented:
rebuild jenkins

timosachsenberg commented:
http://cdash.openms.de/testDetails.php?test=53619633&build=293977

hroest commented:
can you undo 477a9581f5972771e6ea1d4dbee347c1d2d3e1ad ?

Either it should be

```
if (x) {foo();}
```

or 

```
if (x)
{
  foo();
}
```

and not something between.

oliveralka commented:
If I try to do is as follows:
```
        else if (p == IonSource::Polarity::NEGATIVE && feature_charge == 1) {feature_charge = -1;}
        else if (feature_charge == 0) {feature_charge = +1; ++count_to_pos;}
        else if (p == IonSource::Polarity::NEGATIVE) {-abs(feature_charge); ++count_to_neg;}
```
cpplint throws an error:
```
If an else has a brace on one side, it should have it on both  [readability/braces] [5]
```
(That explanes the current style - sorry)
It seems that is a problem with the one liner.

timosachsenberg commented on src/openms/source/ANALYSIS/ID/SiriusMSConverter.cpp:
```diff
@@ -169,16 +169,42 @@ namespace OpenMS
+        }
+        else if (p == IonSource::Polarity::NEGATIVE)
+        {
+          -abs(int_charge); ++count_to_neg;
```
please don't put multiple expressions in the same line

timosachsenberg commented:
why not the short version I proposed?

´´´
if (feature_charge == 0) { feature_charge = 1; }
if (p == IonSource::Polarity::NEGATIVE) { -abs(feature_charge) }
´´´

oliveralka commented:
Think I got it now ...

oliveralka commented:
rebuild jenkins

oliveralka commented:
rebuild jenkins

hroest commented:
looks good but the SiriusAdapter seem to fail quite randomly on our test infrastructure...

oliveralka commented:
Yes, I am not sure about that - For Windows It may be a memory issue. 

In linux I really don't know what is happening with the test - all tests pass on MacOS 14.10 and scientific linux 7.

timosachsenberg commented:
yes, the machine is unfortunately too weak / has not enough ram for the the many VMs it is running

oliveralka commented:
As discussed with @hroest.
Added a switch for SIRIUS (SiriusAdapter) tests (default: ON).
Added dependencies for one test at a time (hopefully reduces memory consumption)

hroest commented:
we discussed this and one solution is to add CMake dependencies so that only one test can run at a time...lets see whether that helps

timosachsenberg commented:
sounds like a good solution. let's see if that works.

oliveralka commented:
I am not able to reproduce the error in Ubuntu 18.04. But it seems to be consistent over several build attempts. 

SiriusAdapter_6 is a converter test - so it tests the .ms file output for the negative mode. This means that SIRIUS is not called in this test. It seems that the "checkFeatureSpectraNumber" is called correctly (see output), but then either the store or copy process goes wrong. 

This should be shown:
```
Number of features to be processed: 1
Number of additional MS2 spectra to be processed: 45
No MS1 spectrum for this precursor. Occurred 1 times.
0 spectra were skipped due to precursor charge below -1 and above +1.
Mono charge assumed and set to charge 1 with respect to current polarity 1 times.
0 features were skipped due to feature charge below -1 and above +1.
SiriusAdapter was used in converter mode and is terminated after openms preprocessing.
If you would like to run SIRIUS internally please disable the converter mode.
``` 

This is shown right now:
```
Number of features to be processed: 1
Number of additional MS2 spectra to be processed: 45
```

I is not reproducible on scientific linux 7. I will try to reproduce the problem in a VM (on the weekend).

hroest commented:
so what happens between 


Number of additional MS2 spectra to be processed: 45

and 

No MS1 spectrum for this precursor. Occurred 1 times.

?

timosachsenberg commented:
rebuild jenkins

timosachsenberg commented:
timeout on travis.. windows unreleated
